libraries/AP_NavEKF2/AP_NavEKF2_Measurements.cpp

```diff
@@ -222,44 +257,25 @@ void NavEKF2_core::readMagData()
         InitialiseVariablesMag();
     }
 
+    // check for a failed compass and switch if failed for magFailTimeLimit_ms
+    if (maxCount > 1 &&
+        !compass.healthy(magSelectIndex) &&
+        imuSampleTime_ms - lastMagRead_ms > frontend->magFailTimeLimit_ms) {
+        tryChangeCompass();
+    }
     
     // do not accept new compass data faster than 14Hz (nominal rate is 10Hz) to prevent high processor loading
     // because magnetometer fusion is an expensive step and we could overflow the FIFO buffer
-    if (use_compass() && compass.last_update_usec() - lastMagUpdate_us > 70000) {
+    if (use_compass() &&
+        compass.healthy(magSelectIndex) &&
+        compass.last_update_usec(magSelectIndex) - lastMagUpdate_us > 70000) {
         frontend->logging.log_compass = true;
 
         // If the magnetometer has timed out (been rejected too long) we find another magnetometer to use if available
         // Don't do this if we are on the ground because there can be magnetic interference and we need to know if there is a problem
         // before taking off. Don't do this within the first 30 seconds from startup because the yaw error could be due to large yaw gyro bias affsets
         if (magTimeout && (maxCount > 1) && !onGround && imuSampleTime_ms - ekfStartTime_ms > 30000) {
```

Comment:
could we still pass here ? the important condition is magTiemeout, but in case of compass issue it should already have been set to false in line 264
and the condition to set magTimeout to true is after readMagData() and is : 
`imuSampleTime_ms - lastMagRead_ms > frontend->magFailTimeLimit_ms)` that you already check on line 264.

So I don't think we could pass here anymore and you probably want to move check block on https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_NavEKF2/AP_NavEKF2_MagFusion.cpp#L262 into line 261 to avoid redundancy !

Reply:
Do you mean like this?

https://github.com/priseborough/ardupilot/commit/c3e6bdc8c1a10bf9a09b57b800e2564a5487d726

It clarifies the two timeout error conditions and ensures the magTimeout flag is updated before readMagData() is called

Reply:
@tridge I think we do need to clean this logic up in both estimators.

Reply:
@priseborough yep, I've put the same logic in both now